### ISSUE TITLE
Code bug not fixed by merge

### ISSUE THREAD

**Issue by drumadrian** [aws/aws-sdk-js-v3#4446]
### Checkboxes for prior research

- [X] I've gone through [Developer Guide](https://docs.aws.amazon.com/sdk-for-javascript/v3/developer-guide) and [API reference](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest)
- [X] I've checked [AWS Forums](https://forums.aws.amazon.com) and [StackOverflow](https://stackoverflow.com/questions/tagged/aws-sdk-js).
- [X] I've searched for [previous similar issues](https://github.com/aws/aws-sdk-js-v3/issues) and didn't find any solution.

### Describe the bug

Hello team, 

This issue is still present today when using the latest SDK: 
https://github.com/aws/aws-sdk-js-v3/issues/2282

The issue was reviewed as a credentials error, but it is indeed a SDK code error as explained here: 
https://github.com/aws/aws-sdk-js-v3/issues/2282#issuecomment-829297405

The is the merge that addressed the credential error but did not address the code error: 
https://github.com/aws/aws-sdk-js-v3/pull/3892

I created an AWS Support case to help raise the visibility of this issue. 
Case ID 12031001901

Since AWS Support does not offer code support, we gotta help each other out.  🤓 

Cheers, 

Adrian 

```------------------------------------------------------------------------------------------------------------```

I am using the newest version of the SDK and my output to the command: node -v    =>  v18.13.0

### SDK version number

    "@aws-sdk/client-firehose": "^3.272.0",     "@aws-sdk/client-kinesis": "^3.272.0"

### Which JavaScript Runtime is this issue in?

Node.js

### Details of the browser/Node.js/ReactNative version

v18.13.0

### Reproduction Steps

I am using the newest version of the SDK and my output to the command: node -v    =>  v18.13.0

I run my code using this command: 
node code.js

```js
import { FirehoseClient, PutRecordCommand } from "@aws-sdk/client-firehose";
console.log('hello world');
const initialDate = new Date().getTime();
// Create a Firehose client object
const firehoseClient = new FirehoseClient({ region: "us-west-2" });
// initial data object
let synthetic_geotab_data = {
    vehicle_mame : 'beamer',
    vehicle_make : 'bmw',
    time : initialDate,
    epoch_millis : initialDate,
    date : initialDate
};
let string_synthetic_geotab_data = JSON.stringify(synthetic_geotab_data);
let data_buffer = Buffer.from(string_synthetic_geotab_data);

for (let x = 0; x < 1000; x++){
    synthetic_geotab_data = get_data();
    string_synthetic_geotab_data = JSON.stringify(synthetic_geotab_data)
    data_buffer = Buffer.from(string_synthetic_geotab_data);
    // Create a PutRecord command object
    const putRecordCommand = new PutRecordCommand({
    DeliveryStreamName: 'geotab',
    Record: {
        Data: data_buffer
      }    
    });

    // Send the record to the Kinesis data stream
    firehoseClient.send(putRecordCommand, function(err, data) {
    if (err) {
        console.log("Failed to send record:", err);
    } else {
        console.log("Record sent successfully:", data);
    }
    });
};

function get_data() {
    // Get the current time in milliseconds since epoch
    let currentDate = new Date().getTime();
    synthetic_geotab_data = {
        vehicle_mame : 'beamer',
        vehicle_make : 'bmw',
        time : currentDate,
        epoch_millis : currentDate,
        date : currentDate
    };
    return synthetic_geotab_data;
};


```

### Observed Behavior

```console
Failed to send record: {"data":{"vehicle_mame":"beamer","vehicle_make":"bmw","time":1676753845626,"epoch_millis":1676753845626,"date":1676753845626}} 


ValidationException: 2 validation errors detected: Value null at 'data' failed to satisfy constraint: Member must not be null; Value null at 'partitionKey' failed to satisfy constraint: Member must not be null

...<remove private data>...

  '$fault': 'client',
  '$metadata': {
    httpStatusCode: 400,
    requestId: 'fce1fcb0-3131-1e08-a08a-c14932c1b1e9',
    extendedRequestId: 'f10/D0cdolqQopinQGhs1NNc7gME4HtO/4GXI3sMnxHHc/sbhEzkekG1ZdAJcdI9ZQw2ORijWYrZIhqVxQNx4yFegkwnptsZgMQ79neWQt0=',
    cfId: undefined,
    attempts: 1,
    totalRetryDelay: 0
  },
  __type: 'ValidationException'
}

```

### Expected Behavior

```
Record sent successfully: {
  '$metadata': {
    httpStatusCode: 200,
    requestId: 'd9c6ba71-3b51-5062-85ad-dd459f7cb937',
    extendedRequestId: 'xaxN8RoiuJNjrrBQnQk3KMxghjEawDlCuVn0vudybLa5301/UvGim561AsW8g8OU7aIB7F1OJEZbr+n72JCX1IG37VlvQueQ',
    cfId: undefined,
    attempts: 1,
    totalRetryDelay: 0
  },

```

### Possible Solution

https://github.com/aws/aws-sdk-js-v3/issues/2282#issuecomment-829297405

### Additional Information/Context

  

> May The Force Be With You

**Comment by RanVaknin** [comment]
Hi @drumadrian ,

I'm not sure what the issue at hand is. I ran your code on my end and its working successfully:

```console
hello world
Record sent successfully: {
  '$metadata': {
    httpStatusCode: 200,
    requestId: 'REDACTED',
    extendedRequestId: 'REDACTED',
    cfId: undefined,
    attempts: 1,
    totalRetryDelay: 0
  },
  Encrypted: false,
  RecordId: 'REDACTED'
}
Record sent successfully: {
...
...
...
x1000 times
```

Are you saying that if you run the request the following way youll get the mentioned error?

```js
const putRecordCommand = new PutRecordCommand({
    DeliveryStreamName: 'PUT-S3-d8w3e',
    Record: {
        Data: get_data()
    }    
})
```

If yes, I am running into the same issue of
```
Failed to send record: TypeError: Cannot read properties of undefined (reading 'byteLength')
...
```
However it is mentioned that `Data` takes in a  `Uint8Array | undefined ` type

<img width="370" alt="image" src="https://user-images.githubusercontent.com/50976344/221031208-5222f828-7fbc-4813-b8d7-6ef3393c2b6b.png">

I guess Im not sure what needs to be fixed here?

Thanks,
Ran~

**Comment by drumadrian** [comment]
Hi @RanVaknin 

Sorry about the confusion.  I updated my code to avoid the error and I forgot to post the original version that gave me an error. 

I am trying to use the PutRecord API call for Kinesis Firehose: 
https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/clients/client-firehose/classes/firehose.html#putrecord

The documentation used to say `Data`  can be a String.  
I think the documentation has been updated.  

Links: 
https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/clients/client-firehose/classes/putrecordcommand.html
https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/clients/client-firehose/interfaces/putrecordcommandinput.html


The AWS documentation is hard to follow sometimes.  I have been using the Python SDK for years and the JavaScript SDK documentation is harder to follow.  

Here is an example of a developer using a String on line `Data: JSON.stringify({` : 
https://github.com/aws/aws-sdk-js-v3/issues/2282

User [arlyon](https://github.com/arlyon) explains that it should be a Unit8Array 


If the SDK could include code examples it would be really helpful.  

Thank you for reviewing this issue.  It can be closed now.  

@drumadrian 🤓

**Comment by RanVaknin** [comment]
@drumadrian ,

Thanks for the info. Yeah, I agree about your comment about the JS V3 documentation. We are in the process of re-launching our documentation. I'm excited for it!

Let us know if you need any additional help.
Thanks,
Ran~
